backend/msgpack/writer.go

```diff
@@ -240,21 +267,64 @@ func (w *writer) encodeWithLock(
 		return nil
 	}
 
-	// Otherwise we get a new buffer from pool, copy the bytes exceeding the
-	// flush size to it, swap the new buffer with the old one, and flush out
-	// the old buffer.
-	newBufferedEncoder := w.encoderPool.Get()
+	// Otherwise enqueue the current buffer.
+	err = w.enqueueBufferWithLock(encoder, sizeBefore)
+	encoder.Unlock()
+	return err
+}
+
+// enqueueBufferWithLock enqueues the buffer of an encoder onto the
+// writer's instance queue. It copies any bytes past sizeBefore into
+// a new buffer before resetting the encoder with that new buffer.
+func (w *writer) enqueueBufferWithLock(
+	encoder *lockedEncoder,
+	sizeBefore int,
+) error {
+	var (
+		oldBufferedEncoder = encoder.Encoder()
+		newBufferedEncoder = w.encoderPool.Get()
+	)
 	newBufferedEncoder.Reset()
 	if sizeBefore > 0 {
-		data := bufferedEncoder.Bytes()
+		var (
+			data      = oldBufferedEncoder.Bytes()
+			buffer    = oldBufferedEncoder.Buffer()
+			sizeAfter = buffer.Len()
+		)
 		newBufferedEncoder.Buffer().Write(data[sizeBefore:sizeAfter])
 		buffer.Truncate(sizeBefore)
 	}
 	encoder.Reset(newBufferedEncoder)
+
+	// Unlock the encoder before we enqueue the old buffer to ensure other
+	// goroutines have an oppurtunity to encode metrics while large timer
+	// batches are being encoded.
 	encoder.Unlock()
 
-	// Write out the buffered data.
-	return w.queue.Enqueue(bufferedEncoder)
+	err := w.queue.Enqueue(oldBufferedEncoder)
```

Comment:
I'd prefer to move the enqueue logic into its own method instead of doing unlocking and re-locking in this method. You don't need to report metrics while holding the lock. Also feels like a cleaner separation of responsibilities and this way you don't need to lock the encoder just to unlock it immediately after it returns.